### PR TITLE
ddl: write DDL binlog.

### PR DESCRIPTION
**Pull request by coocood**
Pass query string in context, also disallows prepare DDL, so we can be sure that
DDL query do not contains parameter marker.

### CODE REVIEW

**Comment by coocood** [comment]
@GregoryIan @iamxy @zimulala @shenli PTAL

File: context/context.go [hunk 1]
```diff
@@ -42,3 +42,19 @@ type Context interface {
 	// ClearValue clears the value associated with this context for key.
 	ClearValue(key fmt.Stringer)
 }
+
+type basicCtxType int
+
+func (t basicCtxType) String() string {
+	switch t {
+	case 1:
```
Comment (tiancaiamao) [inline context/context.go]: when we need other type of basicCtxType otherwise than 1?

File: ddl/ddl_worker.go [hunk 2]
```diff
@@ -22,30 +22,46 @@ import (
...
+		return errors.Trace(err1)
 	})
 	if err != nil {
 		return errors.Trace(err)
 	}
+	commitVer, err := d.store.CurrentVersion()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	d.writePostDDLBinlog(job.ID, startTS, commitVer.Ver)
```
Comment (tiancaiamao) [inline ddl/ddl_worker.go]: check error?

File: ddl/ddl_worker.go [hunk 3]
```diff
@@ -93,6 +109,38 @@ func (d *ddl) doDDLJob(ctx context.Context, job *model.Job) error {
 	}
 }
 
+func (d *ddl) writePreDDLBinlog(ctx context.Context, jobID int64, startTS uint64) error {
```
Comment (tiancaiamao) [inline ddl/ddl_worker.go]: Can this function be an async implementation? so write bin log and two-phase commit run parallel, latency would be lower.

File: ddl/ddl_worker.go [hunk 4]
```diff
@@ -93,6 +109,38 @@ func (d *ddl) doDDLJob(ctx context.Context, job *model.Job) error {
...
+		Tp:       binlog.BinlogType_PreDDL,
+		DdlJobId: jobID,
+		DdlQuery: []byte(ddlQuery),
+		StartTs:  int64(startTS),
+	}
+	err := binloginfo.WriteBinlog(bin)
+	return errors.Trace(err)
+}
+
+func (d *ddl) writePostDDLBinlog(jobID int64, startTS, commitTS uint64) {
```
Comment (tiancaiamao) [inline ddl/ddl_worker.go]: ditto

File: ddl/ddl_worker.go (hunk 3 above)
Comment (coocood) [inline ddl/ddl_worker.go]: DDL is rarely called, it doesn't worth this optimization.

File: ddl/ddl_worker.go (hunk 2 above)
Comment (coocood) [inline ddl/ddl_worker.go]: This function doesn't returns error.

**Comment by tiancaiamao** [comment]
LGTM

File: context/context.go (hunk 1 above)
Comment (c4pt0r) [inline context/context.go]: Why it's an '1' here? Why not a enum?

File: table/tables/tables.go [hunk 5]
```diff
@@ -369,10 +364,9 @@ func (t *Table) AddRecord(ctx context.Context, r []types.Datum) (recordID int64,
 	}
 	if shouldWriteBinlog(ctx) {
 		mutation := t.getMutation(ctx)
-		bin, err := codec.EncodeValue(nil, r...)
-		if err != nil {
-			return 0, errors.Trace(err)
-		}
+		// prepend handle to the row value
+		handleVal, _ := codec.EncodeValue(nil, types.NewIntDatum(recordID))
```
Comment (c4pt0r) [inline table/tables/tables.go]: why remove the error handling?

Comment (coocood) [inline table/tables/tables.go]: Encoding Integer Datum never returns error.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**